lucene/core/src/java/org/apache/lucene/util/automaton/ByteRunAutomaton.java

```diff
@@ -47,44 +47,39 @@ public boolean run(byte[] s, int offset, int length) {
     return accept[p];
   }
 
-  public CharArrayMatcher asCharArrayMatcher(String label) {
-    return new CharArrayMatcher() {
-      @Override
-      public boolean run(char[] chars, int offset, int length) {
-        int state = 0;
-        final int maxIdx = offset + length;
-        for (int i = offset; i < maxIdx; i++) {
-          final int code = chars[i];
-          int b;
-          // UTF16 to UTF8   (inlined logic from UnicodeUtil.UTF16toUTF8 )
-          if (code < 0x80) {
-            state = step(state, code);
-            if (state == -1) return false;
-          } else if (code < 0x800) {
-            b = (0xC0 | (code >> 6));
-            state = step(state, b);
-            if (state == -1) return false;
-            b = (0x80 | (code & 0x3F));
-            state = step(state, b);
+  /**
+   * Returns a representation of the automaton that matches char[] instead of byte[]
+   */
+  public CharArrayMatcher asCharArrayMatcher() {
+    return (chars, offset, length) -> {
+      int state = 0;
+      final int maxIdx = offset + length;
+      for (int i = offset; i < maxIdx; i++) {
+        final int code = chars[i];
+        int b;
+        // UTF16 to UTF8   (inlined logic from UnicodeUtil.UTF16toUTF8 )
+        if (code < 0x80) {
+          state = step(state, code);
+          if (state == -1) return false;
+        } else if (code < 0x800) {
+          b = (0xC0 | (code >> 6));
+          state = step(state, b);
+          if (state == -1) return false;
+          b = (0x80 | (code & 0x3F));
+          state = step(state, b);
+          if (state == -1) return false;
+        } else {
+          // more complex
+          byte[] utf8Bytes = new byte[4 * (maxIdx - i)];
```

Comment:
We should probably create/use this once instead of for each such char?  I know it was like this before; you needn't touch it.

Reply:
++ let's do that in a followup